app/src/main/java/io/apicurio/registry/store/RegistryStorage.java

```diff
@@ -0,0 +1,78 @@
+import io.apicurio.registry.rest.beans.Rule;
+import io.apicurio.registry.rest.beans.VersionMetaData;
+
+/**
+ * The storage layer for the registry.
+ * @author eric.wittmann@gmail.com
+ */
+public interface RegistryStorage {
+
+    public ArtifactMetaData createArtifact(String artifactId, String content) throws ArtifactAlreadyExistsException, RegistryStorageException;
+    
+    public void deleteArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public StoredArtifact getArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public ArtifactMetaData updateArtifact(String artifactId, String content) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public ArtifactMetaData getArtifactMetaData(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public void updateArtifactMetaData(String artifactId, EditableMetaData metaData) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public List<Rule> getArtifactRules(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public void createArtifactRule(String artifactId, Rule rule) throws ArtifactNotFoundException, ArtifactRuleAlreadyExistsException, RegistryStorageException;
+    
+    public void deleteArtifactRules(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public Rule getArtifactRule(String artifactId, String ruleName) throws ArtifactNotFoundException, RuleNotFoundException, RegistryStorageException;
+    
+    public void deleteArtifactRule(String artifactId, String ruleNam) throws ArtifactNotFoundException, RuleNotFoundException, RegistryStorageException;
+    
+    public List<Long> getArtifactVersions(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public StoredArtifact getArtifactVersion(String artifactId, long version) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
+    
+    public void deleteArtifactVersion(String artifactId, long version) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
+    
+    public VersionMetaData getArtifactVersionMetaData(String artifactId, long version) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
+    
+    public void updateArtifactVersionMetaData(String artifactId, long version, EditableMetaData metaData) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
+    
+    public void deleteArtifactVersionMetaData(String artifactId, long version) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
+    
+    public List<Rule> getGlobalRules() throws RegistryStorageException;
+    
+    public void createGlobalRule(Rule rule) throws RegistryStorageException;
```

Comment:
Throw an exception if the rule already exists.

Reply:
Agreed.